flink-runtime/src/main/java/org/apache/flink/runtime/io/network/netty/CreditBasedClientHandler.java

```diff
@@ -269,4 +315,49 @@ private void decodeBufferOrEvent(RemoteInputChannel inputChannel, NettyMessage.B
 			bufferOrEvent.releaseBuffer();
 		}
 	}
+
+	private void writeAndFlushNextMessageIfPossible(Channel channel) {
+		if (channelError.get() != null || !channel.isWritable()) {
+			return;
+		}
+
+		while (true) {
+			RemoteInputChannel inputChannel = inputChannelsWithCredit.poll();
+
+			// The input channel may be null because of the write callbacks that are executed
+			// after each write, and it is also no need to notify credit for released channel.
+			if (inputChannel == null || inputChannel.isReleased()) {
+				return;
+			}
+
+			AddCredit msg = new AddCredit(
+				inputChannel.getPartitionId(),
+				inputChannel.getAndResetCredit(),
+				inputChannel.getInputChannelId());
+
+			// Write and flush and wait until this is done before
+			// trying to continue with the next input channel.
+			channel.writeAndFlush(msg).addListener(writeListener);
+
+			return;
```

Comment:
`return`?

If this is indeed a bug, please add tests that cover it.  Because it seems that we would have now resource leak if we would trigger `notifyCreditAvailable` multiple times while `inputChannel` is not writable.

Reply:
I think it is not a bug. Considering the similar logic is current `PartitionRequestQueue`, if the channel becomes writable via `channelWritabilityChanged` , it will call `writeAndFlushNextMessageIfPossible` again.

`notifyCreditAvailable` just triggers enqueue the `InputChannel` in the pipeline. And if the channel is not writable currently, this `InputChannel` will still keep in the queue. Until the channel becomes writable, the `InputChannel` will be polled from the queue to send credit to the producer.

Reply:
So what is the point of having this `while (true)` if it always terminates after first iteration? 

I still think this return is a mistake. Let's say

1. `notifyCreditAvailable` is called 4 times, enqueuing 4 `InputChannel`s and calling `writeAndFlushNextMessageIfPossible()` 4 times. However  because `channel.isWritable()` returned true, nothing was executed and `inputChannelsWithCredit` has 4 `inputChannels`
2. channel writability changes,  `writeAndFlushNextMessageIfPossible` is called once, this loop rotates only once, only one `inputChanel` is processed, `inputChannelsWithCredit` still has 3 elements, which are dangling indefinitely?

Reply:
First, if `notifyCreditAvailable` is called 4 times, enqueuing 4 `InputChannel` s, but `writeAndFlushNextMessageIfPossible` will be called only once, not 4 times. Because it will only trigger `writeAndFlushNextMessageIfPossible` when the current queue is empty before enqueuing the `InputChannel`, so only the first `InputChannel` will trigger that.

Second, once `writeAndFlushNextMessageIfPossible` is called, it will loop all the `InputChannel` s in the queue until the channel becomes un-writable. Then the left `InputChannel` s in the queue will be chosen after the channel becomes writable again.

Reply:
Firsts of all sorry I made a mistake in my previous comment: 
> However because `channel.isWritable()` returned true

was supposed to be:
> However because `channel.isWritable()` returned false

> Second, once writeAndFlushNextMessageIfPossible is called, it will loop all the InputChannel s in the queue until the channel becomes un-writable. Then the left InputChannel s in the queue will be chosen after the channel becomes writable again.

Can you point out what code will loop through those `InputChannel`s? Can you please take a careful look at the `writeAndFlushNextMessageIfPossible`, especially both exit points from the `while (true)` loop? This code completely doesn't make sense to me.

```
while (true) { // <<========= ?
	RemoteInputChannel inputChannel = inputChannelsWithCredit.poll();

	// The input channel may be null because of the write callbacks that are executed
	// after each write, and it is also no need to notify credit for released channel.
	if (inputChannel == null || inputChannel.isReleased()) {
		return;
	}
	(...)

	return; // <<========= ?
}
```

Reply:
Sorry, I misunderstood your meaning before.
This loop indeed has some redundancies. I will improve it together with UT fixing.